src/overlay/Peer.cpp

```diff
@@ -630,36 +770,237 @@ Peer::recvMessage(StellarMessage const& stellarMsg)
+                    app.getConfig().PEER_PORT);
+                CLOG_ERROR(Overlay, "Dropping connection with {}: {}", err,
+                           e.what());
+                self->drop("Bad crypto request",
+                           Peer::DropDirection::WE_DROPPED_REMOTE,
+                           Peer::DropMode::IGNORE_WRITE_QUEUE);
+            }
+        },
+        fmt::format(FMT_STRING("{} recvMessage"), cat), type);
+}
+
+void
+Peer::sendSendMore(uint32_t numMessages)
+{
+    ZoneScoped;
+    StellarMessage m;
+    m.type(SEND_MORE);
+    m.sendMoreMessage().numMessages = numMessages;
+    sendMessage(std::move(m));
+}
+
+void
+Peer::recvSendMore(StellarMessage const& msg)
+{
+    ZoneScoped;
+    // Flow control must be "on"
+    auto fc = flowControlEnabled();
+    releaseAssert(fc != Peer::FlowControlState::DONT_KNOW);
+
+    if (msg.sendMoreMessage().numMessages == 0)
+    {
+        drop("unexpected SEND_MORE message",
+             Peer::DropDirection::WE_DROPPED_REMOTE,
+             Peer::DropMode::IGNORE_WRITE_QUEUE);
+        return;
+    }
+
+    CLOG_TRACE(Overlay, "Peer {} sent SEND_MORE {}",
+               mApp.getConfig().toShortString(getPeerID()),
+               msg.sendMoreMessage().numMessages);
+
+    // SEND_MORE means we can free some capacity, and dump the next batch of
+    // messages onto the writing queue
+    if (fc == Peer::FlowControlState::ENABLED)
+    {
+        maybeShedLoadAndSendNextBatch(std::nullopt);
+    }
+}
+
+bool
+Peer::hasReadingCapacity() const
+{
+    return flowControlEnabled() != Peer::FlowControlState::ENABLED ||
+           mCapacity.mTotalCapacity > 0;
+}
+
+Peer::FlowControlState
+Peer::flowControlEnabled() const
+{
+    if (mFlowControlState == Peer::FlowControlState::DONT_KNOW)
+    {
+        return Peer::FlowControlState::DONT_KNOW;
+    }
+    if (!mApp.getConfig().ENABLE_OVERLAY_FLOW_CONTROL)
+    {
+        return Peer::FlowControlState::DISABLED;
+    }
+    return mFlowControlState;
+}
+
+void
+Peer::maybeTrimQueue(MessageType lastEnqueuedMessageType)
+{
+    int dropped = 0;
+    if (lastEnqueuedMessageType == TRANSACTION)
+    {
+        auto& queue = mOutboundQueues[1];
+
+        while (queue.size() > Peer::OUTBOUND_TRANSACTION_DROP_LIMIT)
+        {
+            queue.pop_front();
+            dropped++;
+        }
+    }
+    else if (lastEnqueuedMessageType == SCP_MESSAGE)
+    {
+        auto& queue = mOutboundQueues[0];
+
+        // Iterate over the message queue. If we found any messages for slots we
+        // don't keep in-memory anymore, delete those. Otherwise, compare
+        // messages for the same slot and validator against the latest SCP
+        // message and drop
+        auto minSlotToRemember = mApp.getHerder().getMinLedgerSeqToRemember();
+
+        auto latest = queue.back();
+        bool valueReplaced = false;
+
+        for (auto it = queue.begin(); it != queue.end();)
```

Comment:
`deque` may not be the best choice if we end up doing a lot of random location erase (it only performs well when erasing from the head or tail). I think it depends on the type that you're using, if it ends up wrapping shared_ptrs (because of what I wrote elsewhere), then this should not be a problem.